### PR TITLE
nixosTests.ipv6: port to python

### PR DESCRIPTION
**Pull request by mmilata**
###### Motivation for this change

#72828

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/tests/ipv6.nix [hunk 1]
```diff
@@ -35,51 +35,52 @@ import ./make-test.nix ({ pkgs, lib, ...} : {
...
-          my $ip = (split /[ \/]+/, $machine->succeed("ip -o -6 addr show dev $iface scope $scope"))[3];
-          $machine->log("$scope address on $iface is $ip");
-          return $ip;
-      }
+      def wait_for_address(machine, iface, scope):
+          machine.wait_until_succeeds(
+              f"[ `ip -o -6 addr show dev {iface} scope {scope} | grep -v tentative | wc -l` -ge 1 ]"
+          )
+          output = machine.succeed(f"ip -o -6 addr show dev {iface} scope {scope}")
+          ip = re.split(r"[ /]+", output)[3]
```
Comment (tfc) [inline nixos/tests/ipv6.nix]: wouldn't it be more direct and stable to match out the group of the subtext that looks like an ip (four numbers separated by dots)?

File: nixos/tests/ipv6.nix [hunk 2]
```diff
@@ -35,51 +35,52 @@ import ./make-test.nix ({ pkgs, lib, ...} : {
...
+
+      with subtest("loopback address"):
+          client.succeed("ping -c 1 ::1 >&2")
+          client.fail("ping -c 1 ::2 >&2")
 
-      subtest "loopback address", sub {
-          $client->succeed("ping -c 1 ::1 >&2");
-          $client->fail("ping -c 1 ::2 >&2");
-      };
+      with subtest("local link addressing"):
```
Comment (tfc) [inline nixos/tests/ipv6.nix]:
```suggestion
      with subtest("Local link addresses can be obtained and pinged"):
```

File: nixos/tests/ipv6.nix [hunk 3]
```diff
@@ -35,51 +35,52 @@ import ./make-test.nix ({ pkgs, lib, ...} : {
...
+          client.succeed(f"ping -c 1 {client_ip}%eth1 >&2")
+          client.succeed(f"ping -c 1 {server_ip}%eth1 >&2")
 
-      subtest "local link addressing", sub {
-          my $clientIp = waitForAddress $client, "eth1", "link";
-          my $serverIp = waitForAddress $server, "eth1", "link";
-          $client->succeed("ping -c 1 $clientIp%eth1 >&2");
-          $client->succeed("ping -c 1 $serverIp%eth1 >&2");
-      };
+      with subtest("global addressing"):
```
Comment (tfc) [inline nixos/tests/ipv6.nix]:
```suggestion
      with subtest("Global addresses can be obtained, pinged, and reached via http"):
```

File: nixos/tests/ipv6.nix [hunk 4]
```diff
@@ -35,51 +35,52 @@ import ./make-test.nix ({ pkgs, lib, ...} : {
...
-          my $clientIp = waitForAddress $client, "eth1", "global";
-          my $serverIp = waitForAddress $server, "eth1", "global";
-          $client->succeed("ping -c 1 $clientIp >&2");
-          $client->succeed("ping -c 1 $serverIp >&2");
-          $client->succeed("curl --fail -g http://[$serverIp]");
-          $client->fail("curl --fail -g http://[$clientIp]");
-      };
-      subtest "privacy extensions", sub {
-          my $ip = waitForAddress $client, "eth1", "global temporary";
+      with subtest("privacy extensions"):
```
Comment (tfc) [inline nixos/tests/ipv6.nix]:
```suggestion
      with subtest("Privacy extensions: Global temporary address can be obtained and pinged"):
```

**Comment by mmilata** [comment]
I've changed the test according to suggestions but noticed it randomly failing. It's possible that I broke it during the conversion, looking into this ...

File: nixos/tests/ipv6.nix [hunk 5]
```diff
@@ -35,51 +35,52 @@ import ./make-test.nix ({ pkgs, lib, ...} : {
...
-      subtest "local link addressing", sub {
-          my $clientIp = waitForAddress $client, "eth1", "link";
-          my $serverIp = waitForAddress $server, "eth1", "link";
-          $client->succeed("ping -c 1 $clientIp%eth1 >&2");
-          $client->succeed("ping -c 1 $serverIp%eth1 >&2");
-      };
+      with subtest("Global addresses can be obtained, pinged, and reached via http"):
+          client_ip = wait_for_address(client, "eth1", "global")
+          server_ip = wait_for_address(server, "eth1", "global")
+          client.succeed(f"ping -c 1 {client_ip} >&2")
```
Comment (mmilata) [inline nixos/tests/ipv6.nix]: Nondeterministically fails here:
```
Global addresses can be obtained, pinged, and reached via http
client: waiting for success: [ `ip -o -6 addr show dev eth1 scope global | grep -v tentative | wc -l` -ge 1 ]
(1.30 seconds)
client: must succeed: ip -o -6 addr show dev eth1 scope global
(0.02 seconds)
client: global address on eth1 is fd60:cc69:b537:1:3c07:a443:4fa4:3c57
server: waiting for success: [ `ip -o -6 addr show dev eth1 scope global | grep -v tentative | wc -l` -ge 1 ]
(0.04 seconds)
server: must succeed: ip -o -6 addr show dev eth1 scope global
(0.01 seconds)
server: global address on eth1 is fd60:cc69:b537:1:40a5:827c:a880:d577
client: must succeed: ping -c 1 fd60:cc69:b537:1:3c07:a443:4fa4:3c57 >&2
client # PING fd60:cc69:b537:1:3c07:a443:4fa4:3c57(fd60:cc69:b537:1:3c07:a443:4fa4:3c57) 56 data bytes
client # From fd60:cc69:b537:1:3c07:a443:4fa4:3c57 icmp_seq=1 Destination unreachable: Address unreachable
client #
client # --- fd60:cc69:b537:1:3c07:a443:4fa4:3c57 ping statistics ---
client # 1 packets transmitted, 0 received, +1 errors, 100% packet loss, time 0ms
client #
client: output:
error: command `ping -c 1 fd60:cc69:b537:1:3c07:a443:4fa4:3c57 >&2` failed (exit code 1)
```

**Comment by mmilata** [comment]
The problem was in the original test which sometimes ended up using tentative address. Should be fixed now, PTAL.

cc @tfc @flokli @lheckemann

**Comment by lheckemann** [comment]
@ofborg test ipv6

**Pull request merged**